src/graph/executor/query/TraverseExecutor.cpp

```diff
@@ -267,7 +291,133 @@ Status TraverseExecutor::buildInterimPath(GetNeighborsIter* iter) {
+    reqDs_.clear();
+    std::unordered_map<Value, Paths>& current = paths_.back();
+    size_t mapCnt = 0;
+    for (auto& r : results) {
+      if (!r.ok()) {
+        return r.status();
+      } else {
+        mapCnt += r.value().newPaths.size();
+      }
+    }
+    current.reserve(mapCnt);
+    for (auto& r : results) {
+      auto jobResult = std::move(r).value();
+      pathCnt += jobResult.pathCnt;
+      if (!jobResult.reqDs.rows.empty()) {
+        reqDs_.rows.insert(reqDs_.rows.end(),
+                           std::make_move_iterator(jobResult.reqDs.rows.begin()),
+                           std::make_move_iterator(jobResult.reqDs.rows.end()));
+      }
+      for (auto& kv : jobResult.newPaths) {
+        auto& paths = current[kv.first];
+        paths.insert(paths.end(),
+                     std::make_move_iterator(kv.second.begin()),
+                     std::make_move_iterator(kv.second.end()));
+      }
+    }
+    releasePrevPaths(pathCnt);
+    return Status::OK();
+  };
+
+  return runMultiJobs(std::move(scatter), std::move(gather), iter.get());
+}
+
+StatusOr<JobResult> TraverseExecutor::handleJob(size_t begin,
+                                                size_t end,
+                                                Iterator* iter,
+                                                const std::unordered_map<Value, Paths>& prev) {
+  // Handle edges from begin to end, [begin, end)
+  JobResult jobResult;
+  size_t& pathCnt = jobResult.pathCnt;
+  DataSet& reqDs = jobResult.reqDs;
+  reqDs.colNames = reqDs_.colNames;
+  QueryExpressionContext ctx(ectx_);
+  std::unordered_set<Value> uniqueDst;
+  auto* vFilter = traverse_->vFilter() ? traverse_->vFilter()->clone() : nullptr;
+  auto* eFilter = traverse_->eFilter() ? traverse_->eFilter()->clone() : nullptr;
+  const auto& spaceInfo = qctx()->rctx()->session()->space();
+  std::unordered_map<Value, Paths>& current = jobResult.newPaths;
+  for (; iter->valid() && begin++ < end; iter->next()) {
+    auto& dst = iter->getEdgeProp("*", kDst);
+    if (!SchemaUtil::isValidVid(dst, *(spaceInfo.spaceDesc.vid_type_ref()))) {
+      continue;
+    }
+    if (vFilter != nullptr && currentStep_ == 1) {
+      auto& vFilterVal = vFilter->eval(ctx(iter));
+      if (!vFilterVal.isBool() || !vFilterVal.getBool()) {
+        continue;
+      }
+    }
+    if (eFilter != nullptr) {
+      auto& eFilterVal = eFilter->eval(ctx(iter));
+      if (!eFilterVal.isBool() || !eFilterVal.getBool()) {
+        continue;
+      }
+    }
+    auto srcV = iter->getVertex();
+    auto e = iter->getEdge();
+    // Join on dst = src
+    auto pathToSrcFound = prev.find(srcV.getVertex().vid);
+    if (pathToSrcFound == prev.end()) {
+      return Status::Error("Can't find prev paths.");
+    }
+    const auto& paths = pathToSrcFound->second;
+    for (auto& prevPath : paths) {
+      if (hasSameEdge(prevPath, e.getEdge())) {
+        continue;
+      }
+      if (uniqueDst.emplace(dst).second) {
```

Comment:
this can guarantee the uniqueness within a thread, but the uniqueness between different threads cannot be guaranteed
Is it possible to deduplicate  in gather function one more time